core/services/log_storage.rb

```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
+require_relative '../out'
+
+# Class provides storage of logs
+class LogStorage < Out
+  # @param configuration [Session] configuration object that can silence the output
+  def initialize(configuration)
+    super(configuration)
+    @logs = []
+  end
+
+  def out(string)
+    return if string.nil?
+
+    @logs.push(string)
+  end
+
+  def print_to_stdout
+    @logs.each { |log_line| @stream.puts(log_line) }
+  end
+
+  protected
+
+  def print_line(level, string)
+    return if string.nil?
+
+    timestamp = Time.now.strftime('%Y-%m-%dT%H:%M:%S')
+    @logs.push("#{timestamp} #{level}: #{string}")
```

Comment:
Correct solution would be to add the new protected method to the Out class that is responsible for printing a concrete line and call it from within the print_line method.

This could be called print_raw_line. And that method will only have one single line for both the base class and the subclass.